tuftool/src/create.rs

```diff
@@ -77,202 +63,49 @@ pub(crate) struct CreateArgs {
-                version: self.args.snapshot_version,
-                expires: self.args.snapshot_expires,
-                meta: hashmap! {
-                    "root.json".to_owned() => SnapshotMeta {
-                        hashes: Some(Hashes {
-                            sha256: self.root_digest.digest.to_vec().into(),
-                            _extra: HashMap::new(),
-                        }),
-                        length: Some(self.root_digest.size),
-                        version: self.root_digest.root.version,
-                        _extra: HashMap::new(),
-                    },
-                    "targets.json".to_owned() => SnapshotMeta {
-                        hashes: Some(Hashes {
-                            sha256: targets_sha256.to_vec().into(),
-                            _extra: HashMap::new(),
-                        }),
-                        length: Some(targets_length),
-                        version: self.args.targets_version,
-                        _extra: HashMap::new(),
-                    },
-                },
-                _extra: HashMap::new(),
-            },
-            self.args.snapshot_version,
-            "snapshot.json",
-        )?;
-
-        self.write_metadata(
-            Timestamp {
-                spec_version: crate::SPEC_VERSION.to_owned(),
-                version: self.args.timestamp_version,
-                expires: self.args.timestamp_expires,
-                meta: hashmap! {
-                    "snapshot.json".to_owned() => TimestampMeta {
-                        hashes: Hashes {
-                            sha256: snapshot_sha256.to_vec().into(),
-                            _extra: HashMap::new(),
-                        },
-                        length: snapshot_length,
-                        version: self.args.snapshot_version,
-                        _extra: HashMap::new(),
-                    }
-                },
-                _extra: HashMap::new(),
-            },
-            self.args.timestamp_version,
-            "timestamp.json",
-        )?;
+        let targets = self.build_targets()?;
+        let mut editor =
+            RepositoryEditor::new(&self.root).context(error::EditorCreate { path: &self.root })?;
+
+        editor
+            .add_targets(targets)
+            .context(error::TargetsFromDir { dir: &self.indir })?
+            .targets_version(self.targets_version)
+            .targets_expires(self.targets_expires)
+            .snapshot_version(self.snapshot_version)
+            .snapshot_expires(self.snapshot_expires)
+            .timestamp_version(self.timestamp_version)
+            .timestamp_expires(self.timestamp_expires);
+
+        let signed_repo = editor.sign(&self.keys).context(error::SignRepo)?;
+
+        let metadata_dir = &self.outdir.join("metadata");
+        let targets_dir = &self.outdir.join("targets");
+        signed_repo.write(metadata_dir).context(error::WriteRepo)?;
+        signed_repo
```

Comment:
Should we link targets before we write metadata so the metadata is valid/usable as soon as it's there?

Reply:
I'll change it as you're right, but I can't see there being a race condition of any kind here unless a user is using their local machine as a "real" repo and polling for updates constantly. That being said... we have no idea what a user is going to do. :)